Adding information about the decoding attribute (#27166)
mdn/content: files/en-us/web/api/htmlimageelement/decoding/index.md

{{APIRef}}

The **`decoding`** property of the {{domxref("HTMLImageElement")}} interface provides a hint to the browser as to how it should decode the image. More specifically, whether it should wait for the image to be decoded before presenting other content updates or not.

## Value

A string representing the decoding hint. Possible values are:

- `"sync"`
  - : Decode the image synchronously for atomic presentation with other content.
- `"async"`
  - : Decode the image asynchronously and allow other content to be rendered before this completes.
- `"auto"`

>   - : No preference for the decoding mode; the browser decides what is best for the user. This is the default value, but different browsers have different default values — Chromium defaults to `"sync"`, Firefox defaults to `"async"`, and Safari seems to use different values in different circumstances.

Reviewer: ```suggestion
  - : No preference for the decoding mode; the browser decides what is best for the user. This is the default value, but different browsers have different default values — Firefox defaults to `"async"`, Chromium defaults to `"sync"`  as does Safari except in a small number of circumstances.
```

Author: Updated, but turned into bullets to make it easier to read.

## Usage notes

> In theory, this property provides a hint to the browser as to whether it should perform image decoding along with other tasks in a single step (`"sync"`), or allow other content to be rendered before this completes. (`"async"`). In reality, the `decoding` attribute for APIs is often not that useful, and there are often better ways of handling this.
> 
> For images that are not downloaded before being inserted into the DOM, the browser will likely render the empty image initially, and then handle the image when it is available, independently of the other content anyway.
> 
> For images that are downloaded before being inserted into the DOM, using the {{domxref("HTMLImageElement.decode()")}} method is a better way to assure atomic presentation without holding up other content.

Reviewer: Let's tone this down slightly:

```suggestion
The `decoding` property provides a hint to the browser as to whether it should perform image decoding along with other tasks in a single step (`"sync"`), or allow other content to be rendered before this completes. (`"async"`). In reality, the differences between the two `decoding` values for APIs is often difficult to perceive and where there are differences, there is often a better way.

For images that are inserted into the DOM in the viewport, `"async"` can result in flashes of unstyled content, while `"sync"` can result in small amounts of jank. Using the {{domxref("HTMLImageElement.decode()")}} method is usualy a better way to assure atomic presentation without holding up other content.

For images inserted into the DOM outside of the viewport, the image will often be decoded off the main thread in modern browsers before being scrolled into view and there will be no noticeable difference using either value.
```

Author: I've used your version, with a few tweaks

{{APIRef}}

The **`decoding`** property of the {{domxref("HTMLImageElement")}} interface provides a hint to the browser as to whether it should perform image decoding synchronously or asynchronously.

## Value

A string representing the decoding hint. Possible values are:

- `"sync"`
  - : Decode the image synchronously.
- `"async"`
  - : Decode the image asynchronously.
- `"auto"`
  - : No preference for the decoding mode; the browser decides what is best for the user. This is the default value.

## Usage notes

In theory, this property provides a hint to the browser as to whether it should perform image decoding along with other tasks in a single step (`"sync"`), or do other tasks first and then decode the image (`"async"`).

In reality, `decoding` is not very useful. If we imagine a situation in which you have dynamically loaded an image and then added it to the DOM, the decode time will be so small for most images that in most situations you won't notice any difference. In addition, `decoding` doesn't stop the image from being added to the DOM, so you will probably get a flash of empty image as it finishes downloading.

## Examples

> In the below example, you'll likely get an empty image shown on the page as the image is downloaded. Setting `decoding` won't prevent that.

Author: Implemented basically as-is, but with a few tweaks. I changed the example to use `async` instead of `sync`, as the added sentence seems to imply that's what it should be.

> In the below example, you'll likely get an empty image shown on the page as the image is downloaded. Setting `decoding` won't prevent that.

Author: Ah, on second thoughts, I've changed the first code example back to using `sync`, as your examples basically builds on top of the first one, then I've moved the "using `decoding=async` may avoid holding up other content from displaying if the decoding time is long" note to below everything else, as it makes sense to me to handle the sync case first, then talk about the async case.

```js
const img = new Image();
img.decoding = "sync";
img.src = "img/logo.png";
document.body.appendChild(img);
```

> 

Author: Added

> Instead, you can use the {{domxref("HTMLImageElement.decode()")}} method to solve this problem. It provides a way to asynchronously decode an image, pausing inserting it into the DOM until it is fully downloaded and decoded, thereby avoiding the empty image problem described above. This is particularly useful if you're dynamically swapping an existing image for a new one.

Reviewer: ```suggestion
A better way is to use the {{domxref("HTMLImageElement.decode()")}} method to solve this problem. It provides a way to asynchronously decode an image, delaying inserting it into the DOM until it is fully downloaded and decoded, thereby avoiding the empty image problem described above. This is particularly useful if you're dynamically swapping an existing image for a new one. This also prevents other, unrelated, paints outside of this code from being held up while the image is decoding.
```

Then add this code example (again, can't suggest this so you'll need to add manually):

>```js
>async function loadImage(url, elem) {
>  return new Promise((resolve, reject) => {
>    elem.onload = () => resolve(elem);
>    elem.onerror = reject;
>    elem.src = url;
>  });
>}
>
>const img = new Image();
>await loadImage("img/logo.png", img);
>await img.decode();
>document.body.appendChild(img);
>const p = document.createElement('p');
>p.textContent = "Image is fully loaded!"
>document.body.appendChild(p);
>```
 
For the `async` use case however, the attribute may be more useful.

Author: It seemed weird putting a `decode()` example on the `decoding` page, so I've added it to the `decode()` page instead. We do link to it so the reader can find it.

## Browser compatibility

{{Compat}}

## See also

> - The {{domxref("HTMLImageElement.decode()")}} method
> - The {{htmlelement("img")}} element `decoding` attribute.

Reviewer: ```suggestion
- The {{domxref("HTMLImageElement.decode()")}} method
- The {{htmlelement("img")}} element `decoding` attribute.
- [What does the image decoding attribute actually do?](https://www.tunetheweb.com/blog/what-does-the-image-decoding-attribute-actually-do/) on tunetheweb.com (2023)
- 
```
